contracts/test/_fund.js

```diff
@@ -0,0 +1,151 @@
+const {
+  setBalance,
+  setStorageAt,
+} = require("@nomicfoundation/hardhat-network-helpers");
+const ethrs = require("ethers");
+const addresses = require("../utils/addresses");
+const {
+  parseEther,
+  defaultAbiCoder,
+  keccak256,
+  hexlify,
+  zeroPad,
+  solidityKeccak256,
+} = ethrs.utils;
+const { units } = require("../utils/units");
+const erc20Abi = require("./abi/erc20.json");
+
+/**
+ *
+ * Based on https://blog.euler.finance/brute-force-storage-layout-discovery-in-erc20-contracts-with-hardhat-7ff9342143ed
+ * @export
+ * @param {string} tokenAddress
+ * @return {*}  {Promise<number>}
+ */
+const findBalancesSlot = async (tokenAddress) => {
+  const { ethers, network } = (await import("hardhat")).default;
+
+  const encode = (types, values) => defaultAbiCoder.encode(types, values);
+
+  const account = ethrs.constants.AddressZero;
+  const probeA = encode(["uint"], [1]);
+  const probeB = encode(["uint"], [2]);
+
+  const token = await ethers.getContractAt(erc20Abi, tokenAddress);
+
+  for (let i = 0; i < 100; i += 1) {
+    const probedSlot = keccak256(encode(["address", "uint"], [account, i]));
+    if (probedSlot.startsWith("0x0")) continue;
+    // remove padding for JSON RPC
+    // while (probedSlot.startsWith("0x0")) {
+    //     probedSlot = `0x${probedSlot.slice(3)}`
+    // }
+
+    const prev = await network.provider.send("eth_getStorageAt", [
+      tokenAddress,
+      probedSlot,
+      "latest",
+    ]);
+    // make sure the probe will change the slot value
+    const probe = prev === probeA ? probeB : probeA;
+
+    await setStorageAt(tokenAddress, probedSlot, probe);
+
+    const balance = await token.balanceOf(account);
+
+    // reset to previous value
+    await setStorageAt(tokenAddress, probedSlot, prev);
+
+    if (balance.eq(ethrs.BigNumber.from(probe))) return i;
```

Comment:
it'd be nice to cache the results so we don't need to keep finding the slots of the tokens we have already set balances for

Reply:
thanks this is very good point, and a significant speed increase: https://github.com/OriginProtocol/origin-dollar/pull/1872/commits/960914237f5eaa29fef2dd633a71d148f0631c6b